data/src/main/java/org/apache/iceberg/data/DeleteFilter.java

```diff
@@ -226,8 +226,10 @@ private CloseableIterable<T> applyPosDeletes(CloseableIterable<T> records) {
 
     // if there are fewer deletes than a reasonable number to keep in memory, use a set
     if (posDeletes.stream().mapToLong(DeleteFile::recordCount).sum() < setFilterThreshold) {
-      PositionDeleteIndex positionIndex = Deletes.toPositionIndex(filePath, deletes);
-      Predicate<T> isDeleted = record -> positionIndex.isDeleted(pos(record));
+      if (deleteRowPositions == null) {
+        deleteRowPositions = Deletes.toPositionIndex(filePath, deletes);
+      }
+      Predicate<T> isDeleted = record -> deleteRowPositions.isDeleted(pos(record));
```

Comment:
If `deleteRowPositions` is null, is this predicate at risk of an NPE?

Reply:
No NPE. The return of " Deletes.toPositionIndex(filePath, deletes)" must not be null.